src/backend/common/helpers/matchstats_helper.py

```diff
@@ -8,29 +8,146 @@
+)
+from backend.common.models.keys import TeamId
 from backend.common.models.match import Match
-from backend.common.models.stats import EventMatchStats, StatType
-from backend.common.queries import event_query
-
 
+StatAccessor = Callable[[Match, AllianceColor], float]
 TTeamIdMap = Dict[TeamId, int]
 
 
-class MatchstatsHelper(object):
+class MatchstatsHelper:
+    OPR_ACCESSOR: StatAccessor = lambda match, color: match.alliances[color]["score"]
+    DPR_ACCESSOR: StatAccessor = lambda match, color: match.alliances[OPPONENT[color]][
+        "score"
+    ]
+    CCWM_ACCESSOR: StatAccessor = lambda match, color: (
+        match.alliances[color]["score"] - match.alliances[OPPONENT[color]]["score"]
+    )
+    DEFAULT_COMPONENT_ACCESSOR: Callable[[Component], StatAccessor] = lambda comp: (
+        lambda match, color: match.score_breakdown[color].get(comp, 0)
+    )
+
+    EVERGREEN: Dict[StatType, StatAccessor] = {
+        StatType.OPR: OPR_ACCESSOR,
+        StatType.DPR: DPR_ACCESSOR,
+        StatType.CCWM: CCWM_ACCESSOR,
+    }
+
+    COMPONENTS = {
+        2020: {
+            "Total Power Cell Points": lambda match, color: (
+                match.score_breakdown[color].get("autoCellPoints")
+                + match.score_breakdown[color].get("teleopCellPoints")
+            ),
+        },
+        2019: {
+            "Cargo + Panel Points": lambda match, color: (
+                match.score_breakdown[color].get("cargoPoints")
+                + match.score_breakdown[color].get("hatchPanelPoints")
+            )
+        },
+    }
+
+    @classmethod
+    def calculate_matchstats(
+        cls, matches: List[Match], skip_coprs: bool = False, keyed: bool = True
+    ) -> EventMatchstats:
+        return EventMatchstats(
+            oprs=cls.calculate_oprs(matches, keyed=keyed),
+            dprs=cls.calculate_dprs(matches, keyed=keyed),
+            ccwms=cls.calculate_ccwms(matches, keyed=keyed),
+            coprs={} if skip_coprs else cls.calculate_coprs(matches, keyed=keyed),
+        )
+
+    @classmethod
+    def calculate_oprs(cls, matches: List[Match], keyed: bool = True) -> TeamStatMap:
+        return cls.calculate_stat(matches, cls.OPR_ACCESSOR, keyed=keyed)
+
     @classmethod
-    def build_team_mapping(
+    def calculate_dprs(cls, matches: List[Match], keyed: bool = True) -> TeamStatMap:
+        return cls.calculate_stat(matches, cls.DPR_ACCESSOR, keyed=keyed)
+
+    @classmethod
+    def calculate_ccwms(cls, matches: List[Match], keyed: bool = True) -> TeamStatMap:
+        return cls.calculate_stat(matches, cls.CCWM_ACCESSOR, keyed=keyed)
+
+    @classmethod
+    def calculate_coprs(
+        cls, matches: List[Match], keyed: bool = True
+    ) -> Dict[Component, TeamStatMap]:
+        coprs = OrderedDict()
+
+        # If there is not valid data for COPRs, skip
+        if not (
+            matches is not None
+            and len(matches) > 0
+            and matches[0].score_breakdown is not None
```

Comment:
Can we DRY references to this `matches[0]`? It probably means we have to split this check in to two checks - the first one to make sure that `matches` is non-zero, do the assignment, and then to check that the `score_breakdown is not None`

Reply:
Could you clarify more?

Reply:
Yeah! So we reference `matches[0]` 3 different times in this method. Why not write this as...

```
if not (
    matches is not None
    and len(matches) > 0
):
    return coprs

match = matches[0]
if match.score_breakdown is None:
    return coprs
```

and then later on in the code we can do things like -

```
year = match.year
```
```
for component, value in none_throws(match.score_breakdown):
```

It's not a huge thing - just a DRY thing.

Reply:
Also realizing this conditional is written pretty weird. We can probably write it like -

```
if matches is None or len(matches) == 0:
```

The length check can also be replaced with `not matches`, but I'm unsure if that's clearer or just more confusing